tka/builder.go

```diff
@@ -92,8 +93,62 @@ func (b *UpdateBuilder) SetKeyMeta(keyID tkatype.KeyID, meta map[string]string)
 	return b.mkUpdate(AUM{MessageKind: AUMUpdateKey, Meta: meta, KeyID: keyID})
 }
 
+func (b *UpdateBuilder) generateCheckpoint() error {
+	// Compute the checkpoint state.
+	state := b.a.state
+	for i, update := range b.out {
+		var err error
+		if state, err = state.applyVerifiedAUM(update); err != nil {
+			return fmt.Errorf("applying update %d: %v", i, err)
+		}
+	}
+
+	// Checkpoints cant specify a parent AUM.
+	state.LastAUMHash = nil
+	return b.mkUpdate(AUM{MessageKind: AUMCheckpoint, State: &state})
+}
+
+// checkpointEvery sets how often a checkpoint AUM should be generated.
+const checkpointEvery = 15
+
 // Finalize returns the set of update message to actuate the update.
-func (b *UpdateBuilder) Finalize() ([]AUM, error) {
+func (b *UpdateBuilder) Finalize(storage Chonk) ([]AUM, error) {
+	var (
+		needCheckpoint bool    = true
+		cursor         AUMHash = b.a.Head()
+	)
+	for i := len(b.out); i < checkpointEvery; i++ {
+		aum, err := storage.AUM(cursor)
+		if err != nil {
+			if err == os.ErrNotExist {
+				// The available chain is shorter than the interval to checkpoint at.
+				needCheckpoint = false
+				break
+			}
+			return nil, fmt.Errorf("reading AUM: %v", err)
+		}
+
+		if aum.MessageKind == AUMCheckpoint {
+			needCheckpoint = false
+			break
+		}
+
+		parent, hasParent := aum.Parent()
+		if !hasParent {
+			// We've hit the genesis update, so the chain is shorter than the interval to checkpoint at.
+			needCheckpoint = false
+			break
+		}
+		cursor = parent
+	}
+
+	if needCheckpoint {
+		if err := b.generateCheckpoint(); err != nil {
```

Comment:
Should it also delete all of the earlier intermediate AUMs if it's replacing them with a checkpoint?

Reply:
I'd rather keep all the updates that lead to a new state as a kind of audit record - control will keep it forever but (eventually) clients will compact to a checkpoint. This will also let us determine if a bug in the consensus layer is caused by checkpoints vs applying one of the other AUMs.

Reply:
Okay, should we add some sort of verification that requires a checkpoint is equivalent to the state from evaluating everything since the last checkpoint (which, by induction, means it validates back to the beginning of time)?

Reply:
I think that ends up happening anyway? iirc `tka.Open` walks the chain to build the state for the Authority, and then we `state.applyVerifiedAUM` to compute the state we embed in the checkpoint.

Reply:
I mean more like `tka.Open` requires that each checkpoint matches the existing state at that point.

Reply:
Done, though we should remove that check for GA because we expect checkpoints may contain additional fields from new semantics that old clients dont understand. But a good correctness check in the meantime.